examples/query_decomposition/index.ts

```diff
@@ -0,0 +1,54 @@
+import Instructor from "@/instructor"
+import OpenAI from "openai"
+import { z } from "zod"
+
+const QueryTypeSchema = z.enum(["SINGLE", "MERGE_MULTIPLE_RESPONSES"]);
+
+const QuerySchema = z.object({
+  id: z.number(),
+  question: z.string(),
+  dependencies: z.array(z.number()).optional(),
+  node_type: QueryTypeSchema.default("SINGLE")
+})
+
+const QueryPlanSchema = z.object({
+  query_graph: z.array(QuerySchema)
+})
+
+type QueryPlan = z.infer<typeof QueryPlanSchema>
+
+const oai = new OpenAI({
+  apiKey: process.env.OPENAI_API_KEY ?? undefined,
+  organization: process.env.OPENAI_ORG_ID ?? undefined
+})
+
+const client = Instructor({
+  client: oai,
+  mode: "FUNCTIONS"
+})
+
+const createQueryPlan = async (question: string): Promise<QueryPlan | undefined> => {
+  const queryPlan: QueryPlan = await client.chat.completions.create({
+    messages: [
+      {
+        "role": "system",
+        "content": "You are a world class query planning algorithm capable ofbreaking apart questions into its dependency queries such that the answers can be used to inform the parent question. Do not answer the questions, simply provide a correct compute graph with good specific questions to ask and relevant dependencies. Before you call the function, think step-by-step to get a better understanding of the problem.",
+      },
+      {
+        "role": "user",
+        "content": `Consider: ${question}\nGenerate the correct query plan.`,
+      },
+    ],
+    model: "gpt-4-0613",
+    response_model: QueryPlanSchema,
+    max_retries: 1
+  })
+
+  return queryPlan || undefined
+}
+
+const queryPlan = await createQueryPlan(
+  "What is the difference in populations of Canada and the Jason's home country?"
+)
+
+console.log({ queryPlan })
```

Comment:
please add the output in as a comment so its viewable

Reply:
currently running into this val error, any idea why?

```
ZodError: [
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "object",
    "path": [
      "query_graph"
    ],
    "message": "Expected array, received object"
  }
]
 errors: [
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "object",
    "path": [
      "query_graph"
    ],
    "message": "Expected array, received object"
  }
]
```

Reply:
@roodboi ?

Reply:
merge in test-validation and set debug on?

Reply:
The created function looks good to me

```
function_call: {
    name: "response_model",
  },
  functions: [
    {
      name: "response_model",
      description: undefined,
      paramaters: {
        type: "object",
        properties: {
          query_graph: {
            type: "array",
            items: {
              type: "object",
              properties: {
                id: {
                  type: "number",
                },
                question: {
                  type: "string",
                },
                dependencies: {
                  type: "array",
                  items: [Object ...],
                },
                node_type: {
                  type: "string",
                  enum: [ "SINGLE", "MERGE_MULTIPLE_RESPONSES" ],
                  default: "SINGLE",
                },
              },
              required: [ "id", "question" ],
              additionalProperties: false,
            },
          },
        },
        required: [ "query_graph" ],
        additionalProperties: false,
      },
    }
  ],
```

But GPT's response looks pretty weird to me
```
INSTRUCTOR DEBUG:  Raw completion call response:  {
  name: "calculate_differences",
  dependencies: {
    population_of_canada: {
      name: "get_country_population",
      params: {
        country: "Canada",
      },
    },
    population_of_jason_home_country: {
      name: "get_country_population",
      dependencies: {
        jason_home_country: {
          name: "get_home_country",
          params: {
            person: "Jason",
          },
        },
      },
    },
  },
}
```